multibranch/src/test/java/org/jenkinsci/plugins/workflow/multibranch/GitDirectorySCMNavigator.java

```diff
@@ -0,0 +1,109 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.multibranch;
+
+import hudson.Extension;
+import hudson.Launcher;
+import hudson.Proc;
+import hudson.model.TaskListener;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import jenkins.plugins.git.GitSCMSource;
+import jenkins.scm.api.SCMNavigator;
+import jenkins.scm.api.SCMNavigatorDescriptor;
+import jenkins.scm.api.SCMSourceObserver;
+import org.apache.commons.io.IOUtils;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+/**
+ * Sample provider which scans a directory for Git checkouts.
+ */
+public class GitDirectorySCMNavigator extends SCMNavigator {
+
+    private final String directory;
+
+    @DataBoundConstructor public GitDirectorySCMNavigator(String directory) {
+        this.directory = directory;
+    }
+
+    public String getDirectory() {
+        return directory;
+    }
+
+    @Override public void visitSources(SCMSourceObserver observer) throws IOException, InterruptedException {
+        TaskListener listener = observer.getListener();
+        File[] kids = new File(directory).listFiles();
+        if (kids == null) {
+            listener.error(directory + " does not exist");
+            return;
+        }
+        for (File kid : kids) {
+            if (!new File(kid, ".git").isDirectory()) {
+                listener.getLogger().format("Ignoring %s since it does not look like a Git checkout%n", kid);
+                continue;
+            }
+            listener.getLogger().format("Considering: %s%n", kid);
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            Proc proc = new Launcher.LocalLauncher(listener).launch().pwd(kid).cmds("git", "remote", "-v").stdout(baos).start();
+            if (proc.join() != 0) {
+                listener.error("git-remote failed");
+                continue;
+            }
+            String origin = kid.getAbsolutePath(); // fallback
+            for (String line : IOUtils.readLines(new ByteArrayInputStream(baos.toByteArray()))) {
+                Matcher m = ORIGIN.matcher(line);
+                if (m.matches()) {
+                    origin = m.group(1);
+                    listener.getLogger().format("Found origin URL: %s%n", origin);
+                    if (origin.startsWith("git@")) {
+                        listener.getLogger().format("Warning: SSH private keys not yet supported here%n");
```

Comment:
:bug:  You may need to check other users. Otherwise the code may blow up on private instances